src/test/java/com/rog/EShop/controllers/KeycloakControllerTest.java

```diff
@@ -0,0 +1,46 @@
+package com.rog.EShop.controllers;
+
+import com.rog.EShop.dto.TokenDto;
+import com.rog.EShop.properties.KeycloakProperties;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.http.MediaType;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.setup.MockMvcBuilders;
+import org.springframework.web.context.WebApplicationContext;
+
+
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
+import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;
+
+
+@SpringBootTest
+class KeycloakControllerTest {
+    @Autowired
+    private WebApplicationContext wac;
+    private MockMvc mockMvc;
+    private KeycloakProperties keycloakProperties;
+
+    @BeforeEach
+    public void setUp() {
+        this.mockMvc = MockMvcBuilders.webAppContextSetup(this.wac).build();
+    }
+
+    @Test
+    void getToken() throws Exception {
+        TokenDto tokenDto = new TokenDto();
+        tokenDto.setUsername("user5");
+        tokenDto.setPassword("user5");
+        tokenDto.setClientId(keycloakProperties.getClientId());
+        tokenDto.setClientSecret(keycloakProperties.getClientSecret());
+        tokenDto.setGrantType("password");
+        this.mockMvc.perform(post( "/api/token").param(tokenDto.getUsername()))
```

Comment:
unit test is not working. You pass json in wrong way.

```
    @Autowired
    private KeycloakProperties keycloakProperties;

    @Test
    void getToken() throws Exception {
        TokenDto tokenDto = new TokenDto();
        tokenDto.setUsername("admin");
        tokenDto.setPassword("admin");
        tokenDto.setClientId(keycloakProperties.getClientId());
        tokenDto.setClientSecret(keycloakProperties.getClientSecret());
        tokenDto.setGrantType("password");

        ObjectMapper objectMapper = new ObjectMapper();
        ObjectWriter objectWriter = objectMapper.writer().withDefaultPrettyPrinter();
        String requestJson = objectWriter.writeValueAsString(tokenDto);

        this.mockMvc.perform(post("/api/token")
                        .contentType(MediaType.APPLICATION_JSON)
                        .characterEncoding(StandardCharsets.UTF_8)
                        .content(requestJson))
                .andDo(print())
                .andExpect(status().isOk())
                .andExpect(content().contentType(MediaType.APPLICATION_JSON))
                .andExpect(jsonPath("$.access_token").isNotEmpty());
    }
```

For this unit test to work, you need to change getToken() function to next: (add **block()** to the end and change return type of function)
```
    @PostMapping("/token")
    public ResponseTokenDto getToken(@RequestBody TokenDto tokenDto) {
        MultiValueMap<String, String> formData = new LinkedMultiValueMap<>();
        formData.add("username", tokenDto.getUsername());
        formData.add("password", tokenDto.getPassword());
        formData.add("client_id", tokenDto.getClientId());
        formData.add("client_secret", tokenDto.getClientSecret());
        formData.add("grant_type", tokenDto.getGrantType());

        return webClient.post()
                .uri("/realms/ESTORE/protocol/openid-connect/token")
                .header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_FORM_URLENCODED_VALUE)
                .body(BodyInserters.fromFormData(formData))
                .retrieve()
                .bodyToMono(ResponseTokenDto.class)
                .block();
    }
```